### PR TITLE
feature/repo-fetcher

### LINKED ISSUE #10: Codebase fetcher
Implement python module to download codebases from Github and similar.
Should store repository in a specified location.

### LINKED ISSUE #18: Hook PyTypes into testing of other projects
We need to track the executions of other projects from our own project.
Perhaps this will be annotation based, or our project will execute the tests directly.

### LINKED ISSUE #44: Remove pytypes.toml from main directory
The execution of main() fails if there is no ./PyTypes/pytypes.toml file.

Unless this is intended, fix the error and remove the toml file.

### PR DESCRIPTION
**Pull request by bengsparks**
Closes #10 - fetching/repo.py provides a factory function for instantiating the correct class based upon the provided URL. One can also explicitly pass a flag that indicates the type of resource provided at the URL.

Closes #18 - fetching/strat.py provides the basic infra for this. The implementation that is currently in tracing/decorator_appender.py needs to be moved into here, this can be done in a refactor PR during this sprint.

Closes #44 - `pytypes.toml` is not meant to be used during the execution of our project, only in test code, and this PR replaces the `main.py` so that `pytypes.toml` is never needed.

This PR is still a WIP as tests are missing, but I would like feedback on the API before commencing.

### CODE REVIEW

File: fetching/detector.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+
+class TestDetector(ABC):
+    """
+    When given the path to a repo, attempts to detect a specific
+    testing suite and create the appropriate application strategy
+    if there is a match.
+    """
+
+    def __init__(self, project: Project):
```
Comment (RecurvedBow) [inline fetching/detector.py:18]: None-check is missing (or does mypy already handle this?)

File: fetching/__init__.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        params["format"],
+        params["output"],
+        params["verbose"],
+    )
+    logging.basicConfig(level=verb)
+
+    repo = repository_factory(project_url=url, fmt=fmt)
+    project = repo.fetch(out)
+
+    detector = detector_factory(proj=project)
```
Comment (RecurvedBow) [inline fetching/__init__.py:56]: There is no exception handling if detector_factory raises an error. Is this intended?

File: fetching/repo.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        td = self._fetch()
+        self._write_to(td, output)
+
+        return Project(output)
+
+    @property
+    @abstractmethod
+    def fmt(self) -> str:
+        """
+        Static attribute indicating format of repository
```
Comment (RecurvedBow) [inline fetching/repo.py:34]: Static is the wrong word. Use something else.

File: fetching/__init__.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+def main(**params):
+    url, fmt, out, verb = (
+        params["url"],
+        params["format"],
+        params["output"],
+        params["verbose"],
+    )
+    logging.basicConfig(level=verb)
+
+    repo = repository_factory(project_url=url, fmt=fmt)
```
Comment (RecurvedBow) [inline fetching/__init__.py:53]: No exception handling there too.

File: fetching/detector.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        pyproj_cfg = toml.load(pyproj.open())
+        return (
+            "tool" in pyproj_cfg
+            and "poetry" in pyproj_cfg["tool"]
+            and "dev-dependencies" in pyproj_cfg["tool"]["poetry"]
+            and "pytest" in pyproj_cfg["tool"]["poetry"]["dev-dependencies"]
+        )
+
+
+def detector_factory(proj: Project) -> TestDetector:
```
Comment (RecurvedBow) [inline fetching/detector.py:62]: This function could be turned into a static method of TestDetector.

File: fetching/repo.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+    fmt = "Archive"
+
+    def _fetch(self) -> tempfile.TemporaryDirectory:
+        td = tempfile.TemporaryDirectory()
+        git.Repo.clone_from(self.project_url, td.name, depth=1)
+        return td
+
+
+def repository_factory(project_url: str, fmt: str | None) -> Repository:
```
Comment (RecurvedBow) [inline fetching/repo.py:103]: Also here

**Comment by RecurvedBow** [comment]
This looks pretty good. There are some small details, but I think big changes aren't needed.

**Comment by bengsparks** [comment]
What about the ApplicationStrategy? What abstract methods should this class offer? Should implementers work on a project level or on a file level?

File: fetching/detector.py (hunk 1 above)
Comment (bengsparks) [reply fetching/detector.py:18]: mypy handles this for us ^^

File: fetching/__init__.py (hunk 2 above)
Comment (bengsparks) [reply fetching/__init__.py:56]: yeah, there is no way to continue execution if either `repository_factory` or `detector_factory` fail

File: fetching/detector.py (hunk 5 above)
Comment (bengsparks) [reply fetching/detector.py:62]: Interesting idea

File: fetching/repo.py (hunk 6 above)
Comment (bengsparks) [reply fetching/repo.py]: Both are now handled in the new commits.

File: fetching/repo.py (hunk 3 above)
Comment (bengsparks) [reply fetching/repo.py]: Updated.

File: fetching/detector.py (hunk 1 above)
Comment (RecurvedBow) [reply fetching/detector.py:17]: I just checked and sadly, it doesn't :/
Please add a check.

**Review by RecurvedBow — COMMENTED** [review]
There aren't any tests - will they be implemented later?

**Comment by bengsparks** [comment]
> There aren't any tests - will they be implemented later?

yep

File: fetching/detector.py (hunk 1 above)
Comment (bengsparks) [reply fetching/detector.py:17]:
```bash
λ poetry run mypy .
fetching/projio.py:21: error: Argument 1 to "Project" has incompatible type "None"; expected "Path"
```

Could it be you tried to do this from an `__init__.py` or from a file in the testing folders?

**Comment by bengsparks** [comment]
ArchiveRepository should be tested in a separate branch; Ive opened an issue for it at #50. It is important to merge this soon, as much work depends on this new feature

File: tests/fetching/test_gitrepo.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
+import pytest
+
+import fetching
+
+import repoutils
+
+
+# TODO: Convert to setup / teardown
+def test_numpy():
```
Comment (RecurvedBow) [inline tests/fetching/test_gitrepo.py:9]: The test function names do not describe their functionality enough. Use something like if_numpy_repository_is_fetched_repository_is...

**Review by RecurvedBow — CHANGES_REQUESTED** [review]
The test_archive.py file is empty.

File: fetching/detector.py (hunk 1 above)
Comment (RecurvedBow) [reply fetching/detector.py:17]: Oh, I see. Yeah, I tried it in a test file.

**Comment by RecurvedBow** [comment]
Ah, I just saw the issue. Alright, I think after renaming the test function names, this can be merged.

File: tests/fetching/test_gitrepo.py (hunk 7 above)
Comment (bengsparks) [reply tests/fetching/test_gitrepo.py:9]: Good point :+1:

**Comment by bengsparks** [comment]
There seems to be some poetry bug with the old virtualenv. Ive changed how the caching works, and all the tests pass, but we should keep an eye on this.
@RecurvedBow time2merge?

**Review by RecurvedBow — APPROVED** [review]

**Pull request merged**